src/szk.rs

```diff
@@ -0,0 +1,565 @@
+                ..
+            } => Cow::Owned(self.derive_helper_proof_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, jr_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, peer_joint_rand_part) = match proof_share {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    blind_and_joint_rand_part,
+                } => blind_and_joint_rand_part.expect("jr_len > 0 so there is joint_randomness"),
+                SzkProofShare::Helper {
+                    proof_share_seed_and_blind,
+                    joint_rand_part,
+                } => (
+                    proof_share_seed_and_blind,
+                    joint_rand_part.expect("jr_len > 0 so there is joint randomness"),
+                ),
+            };
+            let host_joint_rand_part = self.derive_joint_rand_part(&seed, input_share, nonce)?;
+            let (leader_joint_rand_part, helper_joint_rand_part) = if is_leader {
+                (&host_joint_rand_part, &peer_joint_rand_part)
+            } else {
+                (&peer_joint_rand_part, &host_joint_rand_part)
+            };
+            let (joint_rand_seed, joint_rand) =
+                self.derive_joint_rand(leader_joint_rand_part, helper_joint_rand_part);
+            (joint_rand, Some([joint_rand_seed, host_joint_rand_part]))
+        } else {
+            (Vec::new(), None)
+        };
+        let verifier_share = self.typ.query(
+            input_share,
+            flp_proof_share.as_ref(),
+            &query_rand,
+            &joint_rand,
+            2,
+        )?;
+        Ok(SzkQueryShare {
+            joint_rand_seed_and_part: jr_seed_and_part,
+            verifier: verifier_share,
+        })
+    }
+
+    /// Returns true if the verifier message indicates that the input from which
+    /// it was generated is valid.
+    fn decide(
+        &self,
+        verifier: &[T::Field],
+        l_jr_part: Option<Seed<SEED_SIZE>>,
+        h_jr_part: Option<Seed<SEED_SIZE>>,
+        joint_rand_seed_opt: Option<Seed<SEED_SIZE>>,
+    ) -> Result<bool, SzkError> {
+        // Check if underlying FLP proof validates
+        let check_flp_proof = self.typ.decide(verifier)?;
+        if !check_flp_proof {
+            return Ok(false);
+        }
+        // Check that joint randomness was properly derived from both
+        // aggregators' parts
+
+        match (joint_rand_seed_opt, l_jr_part, h_jr_part) {
+            (Some(joint_rand_seed), Some(leader_joint_rand_part), Some(helper_joint_rand_part)) => {
+                if joint_rand_seed
+                    != self.derive_joint_rand_seed(&leader_joint_rand_part, &helper_joint_rand_part)
+                {
+                    Ok(false)
+                } else {
+                    Ok(true)
+                }
+            }
+            (None, None, None) => Ok(true),
+            (_, _, _) => Ok(false),
```

Comment:
This should never happen. I would return an `Err()`. I think it may even be appropriate to panic here.
```suggestion
            (_, _, _) => panic!("at least one the input seeds is missing"),
```